syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.katalon.px";
option java_outer_classname = "BrowserProto";

package px;

service Browser {
    rpc Do(DoRequest) returns (DoResponse) {}
}

message DoResponse {
    string result = 1;
}

message DoRequest {
    repeated Action actions = 1;
}

message Action {
    oneof action_oneof {
        LaunchAction launchAction = 1;
        CloseAction closeAction = 7;
        GotoAction gotoAction = 2;
        ClickAction clickAction = 3;
        TypeAction typeAction = 4;
        ClearAndTypeAction clearAndTypeAction = 12;
        SelectAction selectAction = 5;
        GetInnerTextAction getInnerTextAction = 6;
        KeyPressAction keyPressAction = 8;
        KeyDownAction keyDownAction = 9;
        KeyUpAction keyUpAction = 10;
        FocusAction focusAction = 11;
    }
}

message LaunchAction {
    bool headless = 1;
}

message CloseAction {
}

message GotoAction {
    string url = 1; // https://www.google.com
}

message ClickAction {
    string selector = 1;
}

message TypeAction {
    string selector = 1;
    string text = 2;
}

message ClearAndTypeAction {
    string selector = 1;
    string text = 2;
}

message SelectAction {
    string selector = 1;
    repeated string values = 2;
}

message GetInnerTextAction {
    string selector = 1;
}

message KeyPressAction {
    string key = 1; // https://github.com/puppeteer/puppeteer/blob/master/lib/USKeyboardLayout.js
}

message KeyDownAction {
    string key = 1; // https://github.com/puppeteer/puppeteer/blob/master/lib/USKeyboardLayout.js
}

message KeyUpAction {
    string key = 1; // https://github.com/puppeteer/puppeteer/blob/master/lib/USKeyboardLayout.js
}

message FocusAction {
    string selector = 1;
}